File: c-sharp/ApiTests/AccessTokenHandlerTest.cs
using System.Net.Http;
using System.Threading.Tasks;
using FluentAssertions;
using Npb.Agview.Api.Example;
using Xunit;

namespace Npb.Agview.ApiTests.Example
{
    public class AccessTokenHandlerTest
    {
        private readonly AccessTokenHandler _sut;
        private readonly HttpClient _httpClient = new();
        private readonly string BaseUrl = Constants.BaseUrl;
        private readonly string ApiKey = Constants.ApiKey;
        private readonly string ApiSecret = Constants.ApiSecret;
        private const int TenSeconds = 10;
        private const int TenThousandSeconds = 10000;

        public AccessTokenHandlerTest()
        {
            _sut = new AccessTokenHandler(_httpClient, BaseUrl, ApiKey, ApiSecret);
        }

        [Fact]
        public async Task GetsNewAccessToken()
        {
            var actual = await _sut.GetNewAccessToken();

            actual.Access.Should().NotBeNull();
            actual.Exp.Should().BeGreaterThan(0);
        }

        [Fact]
        public async Task ReusesExistingNonexpiredToken()
        {
            var originalAccessToken = await _sut.GetNewAccessToken();

            var actual = await _sut.GetNonExpiredOrNewAccessToken(originalAccessToken, TenSeconds);

            actual.Should().BeSameAs(originalAccessToken);
        }

        [Fact]
        public async Task GetsNewAccessTokenWhenExistingOneIsOutsideOfValidityRange()
        {
            var originalAccessToken = await _sut.GetNewAccessToken();

            var actual = await _sut.GetNonExpiredOrNewAccessToken(originalAccessToken, TenThousandSeconds);

            actual.Should().NotBeSameAs(originalAccessToken);
        }
    }
}

File: java/src/test/java/org/npb/agview/api/example/AccessTokenHandlerTest.java
package org.npb.agview.api.example;

import org.hamcrest.Matchers;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.net.http.HttpClient;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;


class AccessTokenHandlerTest {

    private AccessTokenHandler sut;

    private final String BASE_URL = System.getenv("NPB_BASE_URL");
    private final String API_KEY = System.getenv("NPB_API_KEY");
    private final String API_SECRET = System.getenv("NPB_API_SECRET");
    private final HttpClient httpClient = HttpClient.newHttpClient();
    private final int TEN_SECONDS = 10;
    private final int TEN_THOUSAND_SECONDS = 10000;

    @BeforeEach
    public void setup() {
        sut = new AccessTokenHandler(httpClient, new Arguments(BASE_URL, API_KEY, API_SECRET));
    }

    @Test
    public void getsNewAccessToken() {
        var actual = sut.getNewAccessToken();

        assertThat(actual.getAccess(), is(notNullValue()));
        assertThat(actual.getExp(), is(greaterThan(0L)));
    }

    @Test
    public void reusesExistingNonExpiredAccessToken() {
        var originalAccessToken = sut.getNewAccessToken();

        var actual = sut.getNonExpiredOrNewAccessToken(originalAccessToken, TEN_SECONDS);

        assertThat(actual, Matchers.is(originalAccessToken));
    }

    @Test
    public void getsNewAccessTokenWhenExistingOneIsOutsideOfValidityRange() {
        var originalAccessToken = sut.getNewAccessToken();

        var actual = sut.getNonExpiredOrNewAccessToken(originalAccessToken, TEN_THOUSAND_SECONDS);

        assertThat(actual, is(Matchers.not(originalAccessToken)));
    }

}


